OpenKeychain/src/main/java/org/sufficientlysecure/keychain/securitytoken/SecurityTokenConnection.java

```diff
@@ -304,6 +309,140 @@ public void clearSecureMessaging() {
+                        case (byte)0x03:
+                            // using KDF
+                            break;
+                        default:
+                            throw new CardException("Unknown KDF algorithm!");
+                    }
+                    break;
+                case 0x82:
+                    // hash algorithm
+                    switch (tlv.mV[0]) {
+                        case (byte)0x08: // SHA256
+                            digester = new SHA256Digest();
+                            break;
+                        case (byte)0x0a: // SHA512
+                            digester = new SHA512Digest();
+                            break;
+                        default:
+                            throw new CardException("Unknown hash algorithm!");
+                    }
+                    break;
+                case 0x83:
+                    // iteration count
+                    ByteBuffer buf = ByteBuffer.wrap(tlv.mV);
+                    iterations = buf.getInt();
+                    break;
+                case 0x84:
+                    saltPw1 = tlv.mV;
+                    break;
+                case 0x85:
+                    saltPw2 = tlv.mV;
+                    break;
+                case 0x86:
+                    saltPw3 = tlv.mV;
+                    break;
+                case 0x87:
+                    hash_user = tlv.mV;
+                    break;
+                case 0x88:
+                    hash_admin = tlv.mV;
+                    break;
+            }
+        }
+
+        // select salt based on the specified password type
+        switch (password_type) {
+            case 1:
+                salt = saltPw1;
+                break;
+            case 2:
+                salt = saltPw2;
+                break;
+            case 3:
+                salt = saltPw3;
+                break;
+        }
+
+        if (salt == null) {
+            throw new IOException("Error getting salt!");
+        }
+        if (digester == null) {
+            throw new IOException("Error getting hash function!");
+        }
+        if (iterations == 0) {
+            throw new IOException("Error getting iterations!");
+        }
+
+        // prepare input to hash function
+        byte[] data = new byte[salt.length + pin.length];
+        System.arraycopy(salt, 0, data, 0, salt.length);
+        System.arraycopy(pin, 0, data, salt.length, pin.length);
+
+        // hash data repeatedly
+        int q = iterations / data.length;
+        int r = iterations % data.length;
+        for (int i = 0; i < q; i++) {
+            digester.update(data, 0, data.length);
+        }
+        digester.update(data, 0, r);
+
+        byte[] digest = new byte[digester.getDigestSize()];
+        digester.doFinal(digest, 0);
+
+        // delete secrets from memory
+        Arrays.fill(data, (byte) 0);
+
+        // these are less sensitive, but let's be thorough
+        Arrays.fill(salt, (byte) 0);
```

Comment:
these are available without authentication from the card, so they indeed aren't sensitive. I think we might as well cache the DO once we retrieved it (see above)

Reply:
I've removed those lines.